### PR TITLE
core: force read-only flag on .rodata.* sections

### PR DESCRIPTION
**Pull request by jforissier**
This commit fixes a warning with GCC 8.2 that did not occur with GCC 6.2:

 $ make out/arm-plat-vexpress/core/arch/arm/kernel/user_ta.o
   CHK     out/arm-plat-vexpress/conf.mk
   CHK     out/arm-plat-vexpress/include/generated/conf.h
   CHK     out/arm-plat-vexpress/core/include/generated/asm-defines.h
   CC      out/arm-plat-vexpress/core/arch/arm/kernel/user_ta.o
 {standard input}: Assembler messages:
 {standard input}:4087: Warning: setting incorrect section attributes for .rodata.__unpaged

The message is printed as the assembler processes this code fragment,
generated by the C compiler:

 .section        .rodata.__unpaged,"aw"

The older compiler (GCC 6.2) would generate instead:

 .section        .rodata.__unpaged,"a",%progbits

The problem with .rodata.__unpaged,"aw" is that the "w" (writeable) flag
is not consistent with the section name (.rodata.*), which by convention
is supposed to be read-only.

- The section name (".rodata.__unpaged") is given by our macro:
  __rodata_unpaged.
- The "w" flag is added by GCC, supposedly [1] because it knows that
  the data structure contains pointers. In order for the binary to be
  relocatable, pointers should reside in a writeable section. And sure
  enough, if we remove the __rodata_unpaged qualifier, the structure is
  put into the RW data section, which makes perfect sense:

   .section        .data.rel.ro.local.user_ta_ops,"aw"

To avoid the warning, we need to fix the section flags. The section type
(%progbits) is optional, it is deduced from the section name by default.
%progbits indicates that the section contains data (i.e., is not empty).

Link: [1] https://gcc.gnu.org/ml/gcc/2004-05/msg01016.html
Signed-off-by: Jerome Forissier <jerome.forissier@linaro.org>
Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (QEMU)
Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey960)

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Regarding [1], the generated binary isn't relocatable to there's no need for the compiler to add the "w" flag. I'd call this a bug or perhaps a counter-intuitive feature of the compiler.
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
@jenswi-linaro exactly, when I read this explanation I looked for `-fPIC` but obviously we're not using it. I'll modify the commit message slightly because it's a bit of a mystery why "w" is added...

**Comment by jenswi-linaro** [comment]
Spell/word error in commit message.
`s/bug of counter-intuitive/bug or counter-intuitive/`
Other than that, looks good.

File: core/include/kernel/dt.h [hunk 1]
```diff
@@ -35,7 +35,7 @@ struct dt_driver {
 	const void *driver;
 };
 
-#define __dt_driver __section(".rodata.dtdrv")
+#define __dt_driver __section(".rodata.dtdrv" __SECTION_FLAGS_RODATA)
```
Comment (etienne-lms) [inline core/include/kernel/dt.h]: maybe remove since you defined it in the **compiler.h**.

Comment (jforissier) [reply core/include/kernel/dt.h]: Ooops! It should be defined only once indeed.
Actually, I was not sure where the best place would be. I'd rather keep it in `dt.h`, no need to pollute `compiler.h`. I'll remove it from compiler.h if you don't mind.

**Comment by jforissier** [comment]
Update. Moved `__early_ta`, too, for the same reasons as `__dt_driver`.

**Comment by etienne-lms** [comment]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Pull request merged**